plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+        // pop the stack when done
+        scope (exit)
+            ns_stack = ns_stack[0 .. $ - 1];
+
+        auto ns_visitor = scoped!(CppVisitor!(CppNamespace, ControllerT, ProductT))(ctrl,
+                prod, indent, ns_stack, container);
+
+        // fill the namespace with content from the analyse
+        v.accept(ns_visitor);
+        root.put(ns_visitor.root);
+    }
+
+    override void visit(const(TranslationUnit) v) {
+        import std.algorithm : filter;
+        import cpptooling.analyzer.clang.type : makeLocation;
+
+        mixin(mixinNodeLog!());
+
+        LocationTag tu_loc;
+        () @trusted{ tu_loc = LocationTag(Location(v.cursor.spelling, 0, 0)); }();
+
+        if (tu_loc.kind != LocationTag.Kind.noloc && ctrl.doFile(tu_loc.file,
+                "root " ~ tu_loc.toString)) {
+            prod.putLocation(FileName(tu_loc.file), LocationType.Root);
+        }
+
+        v.accept(this);
+    }
+
+    void toString(Writer)(scope Writer w) @safe const {
+        import std.format : FormatSpec;
+        import std.range.primitives : put;
+
+        auto fmt = FormatSpec!char("%u");
+        fmt.writeUpToNextSpec(w);
+
+        root.toString(w, fmt);
+        put(w, "\n");
+        container.get.toString(w, FormatSpec!char("%s"));
+    }
+
+    override string toString() const {
+        import std.exception : assumeUnique;
+
+        char[] buf;
+        buf.reserve(100);
+        toString((const(char)[] s) { buf ~= s; });
+        auto trustedUnique(T)(T t) @trusted {
+            return assumeUnique(t);
+        }
+
+        return trustedUnique(buf);
+    }
+}
+
+private:
+@safe:
+
+import cpptooling.data.representation : CppRoot, CppClass, CppMethod, CppCtor,
+    CppDtor, CFunction, CppNamespace;
+import cpptooling.data.type : LocationTag, Location;
+import cpptooling.data.symbol.container : Container;
+import dsrcgen.cpp : E;
+
+enum ClassType {
+    Normal,
+    Adapter,
+    Gmock
+}
+
+enum NamespaceType {
+    Normal,
+    TestDoubleSingleton,
+    TestDouble
+}
+
+/** Structurally transform the input to a stub implementation.
+ *
+ * TODO should handle StorageClass like cvariant do.
+ *
+ * Ignoring C functions and globals by ignoring the root ranges funcRange and
+ * globalRange.
+ *
+ * Params:
+ *  ctrl: removes according to directives via ctrl
+ */
+CppT rawFilter(CppT, LookupT)(CppT input, Controller ctrl, Products prod, LookupT lookup) @safe {
+    import std.array : array;
+    import std.algorithm : each, filter, map, filter;
+    import std.range : tee;
+    import application.types : FileName;
+    import cpptooling.data.representation : StorageClass, CppNs, CppNsStack;
+    import cpptooling.generator.utility : filterAnyLocation;
+    import cpptooling.utility : dedup;
+
+    // setup
+    static if (is(CppT == CppRoot)) {
+        auto filtered = CppRoot.make;
+    } else static if (is(CppT == CppNamespace)) {
+        auto filtered = input.dup; //CppNamespace.make(input.name);
```

Comment:
(1) Lets assume that rawFilter is responsible for filtering of data. No transformation.
What happens if a namespace is unconditionally duplicated without analysing its content?